### PR TITLE
Peterson algorithm synchronous directed ring leader election

### PR DESCRIPTION
**Pull request by Dalgerok**

### CODE REVIEW

File: leader/directed_ring/peterson.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+func setStatePeterson(v lib.Node, s statePeterson) {
+	data, _ := json.Marshal(s)
+	v.SetState(data)
+}
+
+func processPeterson(v lib.Node, round int) bool {
+	s := getStatePeterson(v)
+	if s.Status != relay {
+		if s.Status == unknown {
```
Comment (krzysztof-turowski) [inline leader/directed_ring/peterson.go:53]: Please collapse these two ifs into:
if s.Status == unknown {
} else if s.Status == leader {
} else if s.Status == relay {
}

File: leader/directed_ring/peterson.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+				} else if m1.Max >= s.Max && m1.Max >= m2.Max {
+					s.Max = m1.Max
+				} else {
+					s.Status = relay
+				}
+			}
+			setStatePeterson(v, s)
+			if s.Status == nonleader {
+				return true
+			}
```
Comment (krzysztof-turowski) [inline leader/directed_ring/peterson.go:86]: Please remove this and just add "return true " to the respective ifs (line 59 and 72 if I see correctly)

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you very much, it's pretty good, only some small corrections are needed.

File: leader/directed_ring/peterson.go (hunk 1 above)
Comment (Dalgerok) [reply leader/directed_ring/peterson.go]: thank you

File: text/1168102/peterson.tex [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+\subsection*{Complexity analysis}
+It is very simple to see that at most $2n \log n + n$ messages are sent. First one
+notes that there can be at most $\lfloor{}\log n \rfloor{}$ phases to reach one \textit{active} process. This follows from the fact that any \textit{active} process survives a phase only if the \textit{active} process to its left does not survive that phase. Hence of $m$ \textit{active} processes during a phase, at most $\lfloor{} \frac{m}{2} \rfloor{}$ will survive for $m > 1$. During a phase every process sends (and receives) two messages, and on the last phase the lone \textit{active} process sends one message which is relayed around to the maximum for a total of $2n \log n + n$.
```
[the remark this answers is not in the document]
Comment (Dalgerok) [reply text/1168102/peterson.tex]: yes

File: text/1168102/peterson.tex [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+\subsubsection*{Procedures}
+
+Each round starts with each \textit{active} sending message ($tid$, \textit{active}) to right neighbour.
+\paragraph{}
+Afterwards, each \textit{active} awaits to receive a message from left.
+
+Let's denote message received from left as $(m1.ntid, m1.status)$. We need to \textit{check} whether $m1.status$ is \textit{leader} or $m1.ntid$ is equal to node's unique identifier. In first case we need set our state to \textit{non-leader} and pass the message to right. In second case we need to declare this node as leader: set state to \textit{leader} and send message with $(i, \textit{leader})$ to right. If \textit{check} is failed then we await for one more message from left as $(m2.nntid, m2.status)$. Do the same \textit{check} as previously, and if it's failed then there are two cases. In first case $m1.ntid \geq tid \land m1.ntid \geq m2.nntid$ we assign $m1.ntid$ to node's $tid$, otherwise we set the node state to \textit{relay}.
+
+\paragraph{}
```
[the remark this answers is not in the document]
Comment (Dalgerok) [reply text/1168102/peterson.tex]: done

File: leader/directed_ring/peterson.go (hunk 2 above)
Comment (Dalgerok) [reply leader/directed_ring/peterson.go]: leads to adding setStatePeterson(v, s) in lines 59 and 72 too

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much for your contribution

**Pull request merged**